### PR TITLE
Lint: add version consistency check

### LINKED ISSUE #1618: Linter: Make sure each child version is valid relative to its parent version
The initial idea for this issue was from @Elchi3, and further refined through discussion. First, we made our understanding of the version explicit:

* **null** == We don't know if it's been implemented or not.
* **true** == It has been implemented but we don't know when.
* **false** == It has not been implemented.
* _version_ == We know it has been implemented starting with this version.

Next, we defined the details of the check we'd like to implement for child versions.

• If the parent version is **false**, then the child version must be **false**.
• If the parent version is **true**, then the child can be anything (no validation required).
• If the parent version is **null**, then we can at least assert that if the child is either **true** or a _version_, there is a mismatch between the parent and the child, although we don't know which is in error.
• If the parent is a _version_, the child version must be the same or a greater version (where greater is determined by the order of the releases in https://github.com/mdn/browser-compat-data/tree/master/browsers).

### PR DESCRIPTION
**Pull request by vinyldarkscratch**
Unfortunately as time has progressed and more changes have been merged into this repository, a number of inconsistencies have been found.  A huge one that occurred is a subfeature being implemented in a browser before its parent feature, if the parent feature was even implemented in the first place.

This PR is a combination from the efforts of #2463, #2875, and #3051, and also closes #1618.  Kudos to @caugner for writing the first half of this PR!  It offers the following consistency checks:

- A subfeature cannot be supported if its parent feature isn't.
- A subfeature cannot be supported in an earlier version than its parent.
- A parent feature cannot have `version_added == null` if its subfeatures are supported

### CODE REVIEW

**Comment by Elchi3** [comment]
Hey, thanks for coming back to this! This is quite big and will need some serious reviewing. I will see how we can get to this. Thank you for some patience in advance :)

**Comment by vinyldarkscratch** [comment]
Yeah, the more I think about it, the more it would be MUCH easier to separate this PR into multiple pieces, haha -- once I go through @ddbeck's comments, I'll be splitting this off into pieces by root folder.  :wink:

File: api/AnimationEvent.json [hunk 1]
```diff
@@ -308,7 +308,7 @@
               "version_added": "10"
             },
             "opera": {
-              "version_added": "10"
+              "version_added": "12"
```
[the remark this answers is not in the document]
Comment (vinyldarkscratch) [reply api/AnimationEvent.json]: In version 12, `AnimationEvent` simply had a prefix for `o`, but I'm down to bump this up to 12.1!

File: api/CSS.json [hunk 2]
```diff
@@ -63,7 +63,7 @@
             "version_added": null
           },
           "webview_android": {
-            "version_added": "66"
+            "version_added": "37"
```
[the remark this answers is not in the document]
Comment (vinyldarkscratch) [reply api/CSS.json]: Since `webview_android` became Chrome-based, most all of its feature support should be the same as `chrome_android`.  The `supports` subfeature indicates `chrome_android` support, and the `webview_android` block is a copy-paste of that with an adjustment of bumping "28" up to "37".

Of course, I could be wrong in this assumption.  :P

File: api/ConstantSourceNode.json [hunk 3]
```diff
@@ -5,10 +5,10 @@
         "mdn_url": "https://developer.mozilla.org/docs/Web/API/ConstantSourceNode",
         "support": {
           "chrome": {
-            "version_added": false
+            "version_added": true
```
[the remark this answers is not in the document]
Comment (vinyldarkscratch) [reply api/ConstantSourceNode.json]: Seems I did a couple of changes backwards by changing the parent rather than the children, haha -- whoops!

File: api/CustomElementRegistry.json [hunk 4]
```diff
@@ -719,4 +719,4 @@
       }
     }
   }
-}
+}
```
[the remark this answers is not in the document]
Comment (vinyldarkscratch) [reply api/CustomElementRegistry.json]: Not sure why either, fixed!

**Comment by vinyldarkscratch** [comment]
Alright, PR splitting has been split.  This PR is **not ready** until the following PRs have all been merged:

- [ ] #3375
- [ ] #3376
- [ ] #3377
- [ ] #3378
- [ ] #3379
- [ ] #3380
- [ ] #3381
- [ ] #3382

**Comment by jpmedley** [comment]
This might be useful as a guide to what's inconsistent. But I can't support merging these. The API one alone introduces numerous Chrome errors. I don't know if the linter's incorrect or what.

**Comment by vinyldarkscratch** [comment]
I’ve run numerous tests on the linter’s behavior to make sure that it’s functioning as expected.  As much as I’d love to say this is an easy fix, this is going to take some time to get through.  Many of the data files already in the repository have inconsistencies that violate this test, and while most are truthfully issues, there seems to be a couple of special cases (which should be handled through notes, I feel).

**Comment by Elchi3** [comment]
By providing smaller PRs, we might actually get a better picture here, so I think could start with reviewing everything that is not in the API folder for now and then see how we could tackle that one in smaller chunks later. Note that myself and other maintainers or staffs only have that much time and so I'm not promising to get to all of this immediately. Thanks for getting into this, though. It is a nice data quality exploration.

**Comment by jpmedley** [comment]
Even still, these are going to take considerably time to get through. The number of edge cases is unknown. And in some cases a decision will need to be made as to whether the data or the test should altered. 

@Elchi3 What order do you want to review them in?

**Comment by foolip** [comment]
In the sub-PRs like https://github.com/mdn/browser-compat-data/pull/3375, is there a data source or is it just fixing inconsistencies by changing one piece of the data? Taking the first file, Animation.idl, it is changing from 40 to 46 for one API member. Is it just picking the earliest release and using that?

**Comment by vinyldarkscratch** [comment]
The data source is within the files themselves.  What each of the sub-PRs is doing is:

- Setting a sub-feature's `version_added` to `false` if the parent feature does not provide support;
- Adding a `version_removed` to the sub-feature if the parent feature has a `version_removed`; and
- Bumping the `version_added` to match or exceed the `version_added` of its parent feature.

In doing so, it seems that we're coming across some inconsistencies in the data where the simple solution isn't the answer.

**Comment by foolip** [comment]
> it seems that we're coming across some inconsistencies in the data where the simple solution isn't the answer

Indeed. The inconsistency means that some data must be wrong, but it seems to me the only way to make it consistent is by finding out which is correct, which can't be derived from the data itself. Perhaps it's better to have consistent data which could be wrong, I'll leave it to other to make that call, I'm not a reviewer.

**Comment by vinyldarkscratch** [comment]
> Perhaps it's better to have consistent data which could be wrong

Personally, I feel the same way about this.  However, we shouldn’t ever have wrong data, which is particularly why the review is taking so long to get this PR, and its sub-PRs, merged.  So far, there’s only been one special case, though quite frankly, it’s one out of thousands.

@jpmedley’s got the right idea in going through and checking the PRs carefully.  Sometimes it’s not the sub-features that have the inaccurate data, it’s the parent feature.  It’s good that we’re checking it all now, though, rather than simply leaving it to remain inaccurate forever.

**Comment by jpmedley** [comment]
"Adding a version_removed to the sub-feature if the parent feature has a
version_removed;"

I believe this is the only case where the change should be made by the
linter. Automatically change the inconsistencies concerns me greatly. The
changed PRs are large, and they take way more time than usual to review.
Meanwhile, we still need to submit PRs for enhancements, improvements, and
spec changes. This is a recipe for nasty merge conflicts.

So I'd like to propose an alternative plan with multiple stages.

   - Phase 1: Flag discrepancies but DO NOT block merges based on them and
   encourage contributors to fix them whenever they can. This phase should be
   short (3 or 4 weeks?). It's main reason for existing is to keep other work
   from being blocked and avoid the pain of nasty merge conflicts. PRs that
   specifically contain discrepancy fixes should be short so that reviewers
   can get through them quickly.
   -  Phase 2: Flag discrepancies but DO block merges based on them. This
   would start after the number of discrepancies has burned down sufficiently.
   Vinyl's origin PRs can give us a sense of how may we have.

In no case should the linter change the file.
Joe Medley | Technical Writer, Chrome DevRel | jmedley@google.com |
 816-678-7195
*If an API's not documented it doesn't exist.*

**Comment by vinyldarkscratch** [comment]
The linter actually doesn’t do the fixes, that’s in a separate script — and much dumber changes at that.  Aside from the changes of setting any `null` values to `false`, all the PR’s modifications have been manual, particularly me going through and just doing the adjustments described up above.

As for the plan you’re proposing...well, to be honest, that’s what I assumed we were already doing.  The sub-PRs are dumb and have no source data other than itself, and as mentioned, the original data needs review already.  (I just don’t have the time available to review every change, which is why I’m glad you are.)  Once all the existing data is reviewed and it’s all ready to merge, that’s when this one will be ready to merge — the final piece that ensures new PRs will follow said consistency.

**Comment by jpmedley** [comment]
What is your basis for changing null to false?
Joe Medley | Technical Writer, Chrome DevRel | jmedley@google.com |
 816-678-7195
*If an API's not documented it doesn't exist.*

**Comment by vinyldarkscratch** [comment]
It was a part of @caugner's original PR that led to this one:

> I have added a fixer now that sets `version_added: false` for a browser in a sub-feature if (and only if) the parent feature is not supported by the browser and the sub-feature currently has `version_added: null` or no entry for this browser.

**Comment by jpmedley** [comment]
That sounds reasonable and safe to me.
Joe Medley | Technical Writer, Chrome DevRel | jmedley@google.com |
 816-678-7195
*If an API's not documented it doesn't exist.*

**Comment by Elchi3** [comment]
> Perhaps it's better to have consistent data which could be wrong,

Yeah, the idea here is to get the consistency ensured by the linter so that we won't introduce these inconsistencies anymore going forward. Of course, to make the CI pass, we now need to fix data we already have.

**Comment by jpmedley** [comment]
That's why I suggested a grace period. It gives us more flexibility in how
we fix the data we already have.
Joe Medley | Technical Writer, Chrome DevRel | jmedley@google.com |
 816-678-7195
*If an API's not documented it doesn't exist.*

**Comment by vinyldarkscratch** [comment]
We still need to figure out what to do with that `implemented_on_audio_context` subfeature in `api/BaseAudioContext.json`, as leaving it as it is will cause the linter to raise an error all the time.  As originally determined, it's a special case feature, however it's the **only** one.  It feels awkward to program an exception for one file out of thousands, so I feel more inclined to include the data in the form of notes.

@Elchi3 @ddbeck Thoughts?

**Comment by jpmedley** [comment]
Since we have [other base classes](https://developer.mozilla.org/en-US/docs/Web/API/Sensor) and could have more in the future this can and probably will happen again. 

Perhaps the relationship should be captured somehow in the BCD itself. Not only would this allow a test to skip over it, it would also allow consumers of the data to show this information on subclasses. (Necessary because base classes can't be instantiated.)

@Elchi3

**Comment by foolip** [comment]
Recording the parent interface of each interface would be neat, and is something that I could extract from the IDL in reffy-reports.

**Comment by Elchi3** [comment]
Interesting idea, can you add that to https://github.com/mdn/browser-compat-data/issues/472 or maybe start a new issue where we can discuss how to represent "parent interfaces" in the data model?

**Comment by jpmedley** [comment]
Done: #3441

File: api/AudioContext.json [hunk 5]
```diff
@@ -405,6 +405,82 @@
...
+          "mdn_url": "https://developer.mozilla.org/docs/Web/API/BaseAudioContext/createConstantSource",
+          "support": {
+            "chrome": {
+              "version_added": "10",
+              "version_removed": "56",
+              "prefix": "webkit",
+              "notes": "Available as a part of <code>BaseAudioContext</code>."
+            },
+            "chrome_android": {
+              "version_added": "33",
```
Comment (jpmedley) [inline api/AudioContext.json]: Do you happen to know where 33 came from? This should be 18.

File: api/HTMLCanvasElement.json [hunk 6]
```diff
@@ -52,159 +52,6 @@
           "deprecated": false
         }
       },
-      "height": {
-        "__compat": {
-          "mdn_url": "https://developer.mozilla.org/docs/Web/API/HTMLCanvasElement/height",
-          "support": {
-            "chrome": {
-              "version_added": "4"
```
Comment (jpmedley) [inline api/HTMLCanvasElement.json]: Chrome still supports height. Why is this being removed?

File: api/HTMLCanvasElement.json [hunk 7]
```diff
@@ -52,159 +52,6 @@
...
-            "deprecated": false
-          }
-        }
-      },
-      "width": {
-        "__compat": {
-          "mdn_url": "https://developer.mozilla.org/docs/Web/API/HTMLCanvasElement/width",
-          "support": {
-            "chrome": {
-              "version_added": "4"
```
Comment (jpmedley) [inline api/HTMLCanvasElement.json]: Chrome still supports width. Why is this being removed?

File: api/AudioContext.json (hunk 5 above)
Comment (vinyldarkscratch) [reply api/AudioContext.json]: (This change is a part of #3426.)  The number "33" came from the original PR that added `api.BaseAudioContext`.  If "18" is correct, then I'll push a commit to #3426 that will update it!

Comment (jpmedley) [reply api/AudioContext.json]: Please do.

**Comment by Elchi3** [comment]
fwiw, I don't think it is necessary that you merge PRs into this PR all the time. Actually, I think it would be nice to see travis failing to review just the script and how it is reporting real errors.

**Comment by vinyldarkscratch** [comment]
Alrighty then, leaving it as a failing check it is!  Once all the sub-PRs have been merged, I'll then rebase.  :wink:

**Comment by vinyldarkscratch** [comment]
Today is now the day!  Time for the last step before it's ready for merging!

**Comment by vinyldarkscratch** [comment]
Looks like as time went on, we ended up getting a couple of inconsistencies in the CSS data -- fixes coming right up!

**Comment by ddbeck** [comment]
@vinyldarkscratch whoa, I didn't realize we had merged all the other PRs! This has been a long time coming. Way to go! 🎊 

That said, I wanted to warn you that it'll be a few days to get this merged, even with the fixes to get the linter to pass. Historically, I haven't merged linter changes without @Elchi3's approval and he's on holiday until next week.

Thank you again!

**Comment by vinyldarkscratch** [comment]
Haha, not a problem, I totally understand!  I hope they're having a wonderful vacation.  :wink:  And hey, we'll have a wonderful gift awaiting their return!

**Comment by vinyldarkscratch** [comment]
Haha, not a problem, I totally understand!  I hope they're having a wonderful vacation.  :wink:  And hey, we'll have a wonderful gift awaiting their return!

Speaking of which, I just committed some fixes to all the existing files; all except for one.  @ddbeck, you remember #3587?  It looks like we've got some inconsistencies in there.  Unfortunately, it seems that HTML support was implemented earlier than SVG support (at least by what the data says), but Edge/IE only have SVG support, no HTML support.  Either way, the linter will be yelling at us to fix it.  I'm thinking that we should have a subfeature for SVG and HTML each, and then just set the "basic support" to whichever versions are earlier.  Thoughts?

**Comment by vinyldarkscratch** [comment]
I opened up #3638 to address this issue and possible solution.

**Comment by ddbeck** [comment]
For the record, @vinyldarkscratch's solution to the clip-path SVG/HTML problem was good. 👍

**Comment by Elchi3** [comment]
Great stuff! Really fantastic work! :) 
In order to best review this, I'd like to see the remaining data fixes moved to a separate PR and also I believe we don't want to add all the "fix" scripts to the repo. Let's focus our reviewing here on test-consistency.js. I will try to get to this shortly.

**Comment by vinyldarkscratch** [comment]
Remaining fixes have been separated into a new PR!  As far as the "fix" scripts, I included them from @caugner's original PR as I thought it would be nice to include a basic fix (and the same file is used in #3427), but I'd be happy to remove it if it's simply not desired?

**Comment by Elchi3** [comment]
I'm just thinking that we won't need those fixes again once we've enforced the stricter rules by tests. It is also less to review then. The fix script for ordering might be less controversial than the fix for consistency, but that one doesn't belong to this PR.

**Comment by vinyldarkscratch** [comment]
Guess you’ve got a point — alright, sure thing, will remove!

**Comment by vinyldarkscratch** [comment]
Argh, we were so close!  The last PR was merged, but there have been more inconsistencies that have popped up.  Well...time to get back to work on this with another PR!

**Comment by vinyldarkscratch** [comment]
I figured out a way to treat ranged values better.  I found some false positives in ranged values during my last run, caused by a ranged value for the parent (≤37) and a child with a more defined value (4.4.3).  This is because I was only removing the range delimiter, which would treat the ranged value as the highest version.  I have now set it so that if a range is present, the consistency treats it as version "1" of the browser.  No more false positives from the ranged values now!

**Comment by vinyldarkscratch** [comment]
Great news -- only two more PRs to get this linter to pass!

**Comment by vinyldarkscratch** [comment]
Great news, this linter finally passes once again, as all consistency has been fixed!  This is now ready for merging!

**Comment by vinyldarkscratch** [comment]
ARGH!  Looks like a few recent PRs created some more inconsistencies...

**Comment by caugner** [comment]
Wouldn't it be possible to freeze the repository temporarily (i.e. not merge other PRs) once everything is consistent again? After that, all other PRs would need to be rebased anyways.

Or, alternatively, couldn't this PR be merged, even if there are few remaining inconsistencies (since merging other unrebased PRs after merging a consistent state might introduce new consistencies - and break the master build - anyway)?

**Comment by vinyldarkscratch** [comment]
Whoo, the linter passed in a local run and we've fixed all the consistency!  This is ready to merge once again!

File: test/linter/test-consistency.js [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+   * @param {array} path
+   * @returns {Array<object>}
+   */
+  checkSubfeatures(data, path = []) {
+    let allErrors = [];
+
+    // Check this feature.
+    if (this.isFeature(data)) {
+      const feature = path.length ? path[path.length - 1] : 'ROOT';
+      const featurePath = path.length ? path.slice(0, path.length - 1).join('.') : '';
```
Comment (Elchi3) [inline test/linter/test-consistency.js]: Is `featurePath` unused?

Comment (vinyldarkscratch) [reply test/linter/test-consistency.js]: It is, removing!

File: test/linter/test-consistency.js [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+  if (errors.length) {
+    const relativeFilename = path.relative(process.cwd(), filename);
+    console.error(chalk`{red   Consistency - }{red.bold ${errors.length} }{red ${errors.length === 1 ? 'error' : 'errors'}:}`);
+    errors.forEach(({ feature, path, errors }) =>  {
+      console.error(chalk`{red   → }{red.bold ${errors.length}}{red  × }{red.bold ${feature}}{red  [${path.join('.')}]: }`);
+      errors.forEach(({ errortype, browser, subfeatures }) => {
+        if (errortype == "unsupported") {
+          console.error(chalk`{red     → No support in }{red.bold ${browser}}{red , but this is not declared for sub-feature(s): }{red.bold ${subfeatures.join(', ')}}`);
+        } else if (errortype == "support_unknown") {
+          console.error(chalk`{red     → No support known in }{red.bold ${browser}}{red , but support declared in the following sub-feature(s): }{red.bold ${subfeatures.join(', ')}}`);
```
[the remark this answers is not in the document]
Comment (vinyldarkscratch) [reply test/linter/test-consistency.js]: I like that idea, yeah -- let me see what I can do for that!

[the remark this answers is not in the document]
Comment (vinyldarkscratch) [reply test/linter/test-consistency.js]:
```
✖ api/AbortController.json
  Consistency - 1 error:
  → 1 × AbortController [api.AbortController]: 
    → No support in chrome, but support is declared in the following sub-feature(s):
      → AbortController: null
      → abort: 66
      → signal: 66
```

[the remark this answers is not in the document]
Comment (Elchi3) [reply test/linter/test-consistency.js]: Yes, I like this better! Thank you :)

**Comment by Elchi3** [comment]
Yes, or back things out or correct things otherwise. I think we would manage. Let's do this! :)

**Comment by vinyldarkscratch** [comment]
Woo, celebration time!

I have to give @caugner a HUGE thanks for this -- without them, this probably wouldn't have come to life.  Thank you so much, @caugner!

**Comment by caugner** [comment]
Congratulations to you, @vinyldarkscratch! :tada:
Thank you for working on this so hard for the last 8,5 months! :birthday:

**Comment by jpmedley** [comment]
Yes indeed.
